src/event.cc

```diff
@@ -1,325 +1,251 @@
 }
 
 //
-// C++ callable function to create a JavaScript Event object.
+// add KV pair
 //
-// This signature includes an optional boolean for whether an edge
-// should be set or not. It defaults to true.
+// returns -1 if error.
 //
-Napi::Object Event::NewInstance(Napi::Env env, oboe_metadata_t* omd, bool edge) {
-  Napi::EscapableHandleScope scope(env);
+int addKvPair(oboe_event_t* event, std::string key, Napi::Value value) {
+  int status = -1;
 
-  Napi::Value v0 = Napi::External<oboe_metadata_t>::New(env, omd);
-  Napi::Value v1 = Napi::Boolean::New(env, edge);
+  if (value.IsBoolean()) {
+    status = oboe_event_add_info_bool(event, key.c_str(), value.As<Napi::Boolean>().Value());
+  } else if (value.IsNumber()) {
+    const double v = value.As<Napi::Number>();
 
-  Napi::Object o = constructor.New({v0, v1});
-
-  return scope.Escape(napi_value(o)).ToObject();
-}
+    // it it's within integer range add an integer
+    if (v > -(pow(2, 53) - 1) && v < pow(2, 53) -1) {
+      status = oboe_event_add_info_int64(event, key.c_str(), v);
+    } else {
+      status = oboe_event_add_info_double(event, key.c_str(), v);
+    }
+  } else if (value.IsString()) {
+    std::string s = value.As<Napi::String>();
 
-//
-// JavaScript callable method to format the event as a string.
-//
-// The string representation of an event is the string representation of
-// the metadata part of the event.
-//
-Napi::Value Event::toString(const Napi::CallbackInfo& info) {
-    oboe_metadata_t* md = &this->event.metadata;
-    char buf[Metadata::fmtBufferSize];
-
-    int rc;
-
-    if (info.Length() == 1 && info[0].IsNumber()) {
-      int fmt = info[0].As<Napi::Number>().Int64Value();
-      // default 1 to a human readable form for historical reasons.
-      if (fmt == 1) {
-        fmt = Metadata::fmtHuman;
-      }
-      rc = Metadata::format(md, sizeof(buf), buf, fmt) ? 0 : -1;
+    // make a stab at detecting binary data
+    if (memchr(s.c_str(), '\0', s.length())) {
+      status = oboe_event_add_info_binary(event, key.c_str(), s.c_str(), s.length());
     } else {
-      rc = oboe_metadata_tostr(md, buf, sizeof(buf) - 1);
+      status = oboe_event_add_info(event, key.c_str(), s.c_str());
     }
+  }
 
-    return Napi::String::New(info.Env(), rc == 0 ? buf : "");
+  return status;
 }
 
 //
-// JavaScript callable method to set the event's sample flag to the boolean
-// sense of the argument.
+// add an edge
 //
-// returns the previous value of the flag.
-//
-Napi::Value Event::setSampleFlagTo(const Napi::CallbackInfo& info) {
-    Napi::Env env = info.Env();
+int addEdge(oboe_event_t* event, Napi::Value edge) {
+  // oboe returns -1 for errors, so this returns an error of 1
+  // so the two causes can be distinguished.
+  int status = 1;
 
-    if (info.Length() != 1) {
-        Napi::Error::New(env, "setSampleFlagTo requires one argument").ThrowAsJavaScriptException();
-        return env.Null();
-    }
+  if (!edge.IsObject() || edge.IsArray()) {
```

Comment:
it must be an object so that check is required. it also must *not* be an array (which is an object) so that check is required too. if `!edge.IsArray()` was the check then edge could be anything other than an array (string, number, boolean), and it must be an object. i could make the checks more stringent, e.g., verify that `Napi::Object mb` is a `Napi::Buffer` but i'm not trying to double check all the javascript code. it pretty much has to be right or the agent's not going to work anyway.